### PR TITLE
Initialize ops module with necessary functions

### PR DESCRIPTION
**Pull request by lsetiawan**
## Overview

This PR adds all the necessary functions for calculations and computations to an `ops` module. This includes the following:

Screenshot 2023-05-25 at 12 48 32 PM

### CODE REVIEW

File: src/gnatss/ops.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    if num_dims == 1:
+        # 1D array
+        return np.array(np.trim_zeros(input_array))
+    elif num_dims == 2:
+        # 2D array
+        return np.array(
+            [np.trim_zeros(arr) for arr in input_array if np.trim_zeros(arr).size > 0]
+        )
+
+    raise ValueError("Only 1 or 2-D arrays are supported")
```
Comment (carlosgjs) [inline src/gnatss/ops.py:375]:
```suggestion
    raise ValueError(f"Only 1 or 2-D arrays are supported, instead for {num_dims} dimensions")
```

File: src/gnatss/ops.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+        If the input vector is not a 1-D array
+    """
+
+    if input_vector.shape != (3,):
+        raise ValueError("Unit vector calculation must be 1-D array of shape 3!")
+
+    vector_norm = np.linalg.norm(input_vector)
+
+    if vector_norm == 0:
+        return np.array([2.0, 0.0, 0.0])
```
Comment (carlosgjs) [inline src/gnatss/ops.py:169]: Could this be defined as a constant instead?

**Review by carlosgjs — COMMENTED** [review]
Minor comments, I'll let John approve since I don't understand any of the calculations :)

File: tests/test_ops.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+def test_calc_lsq_contrained():
+    X, XP, MX, MXP = calc_lsq_contrained(ATWA_SAMPLE, ATWF_SAMPLE)
+
+    # The results are in meters and we'd like to ensure cm
+    # precision, so 1e-4 should be very precise,
+    # might even be overkill
+    assert np.allclose(X, LSQ_RESULT[0], atol=1e-4)
+    assert np.allclose(XP, LSQ_RESULT[1], atol=1e-4)
+    assert np.allclose(MX, LSQ_RESULT[2], atol=1e-4)
+    assert np.allclose(MXP, LSQ_RESULT[3], atol=1e-4)
```
Comment (lsetiawan) [inline tests/test_ops.py:277]: @johnbdesanto Is this an okay precision tolerance to assume equivalents in the data? 1e-4?

Comment (johnbdesanto) [reply tests/test_ops.py:300]: The legacy code reports positioning precise to 1e-4 m, with uncertainties that have more significant figures. Practically speaking, I think this is a good level although the code is probably going more precise under the hood. (From our discussion the other week, the criteria the old code uses is some small fraction of the standard deviation, which can be much smaller than 1e-4.) For reference, here is an excerpt for a sol_out file:

After iteration  3, rms residual =     45.85 cm, error factor =    10.255
 09 TRANSPONDER (23162 MEASUREMENTS):
   x =  -2690793.3252 +/- 1.46558D-04 m   del_e =      0.0466 +/- 1.50678D-04 m
   y =  -3835618.4714 +/- 1.48229D-04 m   del_n =     -0.1612 +/- 1.48716D-04 m
   z =   4310864.2626 +/- 1.49188D-04 m   del_v =         0.2385 +/- 1.44526D-04 m
   Lat. =  42.808288298 deg., Long. = -125.050730740 deg., Hgt.msl = -1499.1265 m
 10 TRANSPONDER (23162 MEASUREMENTS):
   x =  -2690858.9732 +/- 1.46558D-04 m   del_e =      0.0467 +/- 1.50679D-04 m
   y =  -3833685.5512 +/- 1.48229D-04 m   del_n =     -0.1613 +/- 1.48714D-04 m
   z =   4312494.0363 +/- 1.49188D-04 m   del_v =         0.2385 +/- 1.44527D-04 m
   Lat. =  42.828504308 deg., Long. = -125.064967519 deg., Hgt.msl = -1524.3355 m
 11 TRANSPONDER (23162 MEASUREMENTS):
   x =  -2689077.5316 +/- 1.46558D-04 m   del_e =      0.0465 +/- 1.50677D-04 m
   y =  -3834901.8001 +/- 1.48229D-04 m   del_n =     -0.1613 +/- 1.48715D-04 m
   z =   4312566.9845 +/- 1.49188D-04 m   del_v =         0.2385 +/- 1.44528D-04 m
   Lat. =  42.829154027 deg., Long. = -125.038583041 deg., Hgt.msl = -1494.8755 m

**Review by johnbdesanto — APPROVED** [review]
Looks fine as far as I can tell. One thing that stood out to me is that the tolerance for the gps 3dsig value is currently hard-coded to 0.05 m. While I think this is a fine and practical level, we may at some point want to relax this and let the user set a tolerance themselves. Not immediately critical, though.

**Comment by lsetiawan** [comment]
> gps 3dsig value is currently hard-coded to 0.05 m

Yepp. This is just the default value, user is able to specify what they want it to be 😄

**Comment by lsetiawan** [comment]
This PR closes #29, #34, #35, #38, #39, #40, #41

**Pull request merged**